libc/string_cpy.c

```diff
@@ -0,0 +1,1527 @@
+	/* Checking capability of handling big blocks of data */
+	create_big_string();
+	TEST_ASSERT_EQUAL_PTR(buff, memcpy(buff, longStr, sizeof(longStr)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, longStr, sizeof(longStr));
+
+	/*
+	 * Offset big chunk of data remembering we want to
+	 * copy data from point 128 and we must trim length of cpy to it size
+	 */
+	TEST_ASSERT_EQUAL_PTR(&buff[128], memcpy(&buff[128], longStr, sizeof(longStr) - 128));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&buff[128], longStr, sizeof(buff) - 128);
+
+#pragma GCC diagnostic pop
+}
+
+
+/*
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_memccpy)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memccpy)
+{
+}
+
+
+TEST(string_memccpy, basic)
+{
+	/* Ifdef used because lack of function 'memccpy' */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+
+	char strDest[MAX_STR_LEN] = { 0 },
+		 strPlaceholder[MAX_STR_LEN] = { 0 };
+	int i;
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strDest, TEST_STR1, 'x', sizeof(TEST_STR1)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strDest);
+
+	/* Checking if we doesn't overwrite elements after end of input*/
+	for (i = strlen(strDest); i < sizeof(strDest) - strlen(strDest); i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', strDest[i]);
+	}
+
+	/* Copy only half of string str2 and in addition we searching for letter 's' on place 8 where string contains 20 elements*/
+	TEST_ASSERT_EQUAL_PTR(&strPlaceholder[sizeof(TEST_STR2) / 2 - 2], memccpy(strPlaceholder, TEST_STR2, 's', sizeof(TEST_STR2) / 2));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strPlaceholder, TEST_STR2, sizeof(TEST_STR2) / 2 - 2);
+
+#pragma GCC diagnostic pop
+#endif
+}
+
+
+TEST(string_memccpy, stop_char_found)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char bigStrDest[BIG_NUMB] = { 0 };
+	int i;
+
+	create_big_string();
+	/*
+	 * Memccpy will return a byte after found endpoint
+	 * in case of find a smaller dataset it will always point to on next byte in the array
+	 * In this case we use longStr and try to find all chars between [1 and 255]
+	 * and check if the pointer that memccpy return is the same as the point where value should be
+	 */
+	for (i = 1; i < MAX_ASCII_SET; i++) {
+		TEST_ASSERT_EQUAL_PTR(&bigStrDest[i + 1], memccpy(bigStrDest, longStr, longStr[i], MAX_ASCII_SET));
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(bigStrDest, longStr, i);
+	}
+
+#endif
+}
+
+
+TEST(string_memccpy, stop_int_found)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char bigStrDest[BIG_NUMB] = { 0 };
+	/*
+	 * Memccpy accepts int as an element of stop and converts it to uint8
+	 * We use -1 and +1 to MAX_ASCII_SET to not go over with values and stick to 255 and -255
+	 */
+	int bigIntDest[] = { 1, -MAX_ASCII_SET + 1, MAX_ASCII_SET - 1, INT_MAX / 5, INT_MAX / 3, INT_MAX, INT_MIN / 5, INT_MIN / 3, INT_MIN + 1 };
```

Comment:
Our variable longStr itself has null terms. Every time the number from `bigIntDest` converted to `uint8` equals `0` we will always acquire a spot where the first null term is present, in our case it will be a place defined by `MAX_ASCII_SET`. Created case in case of this situation has a place.

A solution in case of this situation takes place will be created and the values of `bigIntDest` could be chosen freely.